zio-http/src/main/scala/zio/http/api/MiddlewareSpec.scala

```diff
@@ -57,11 +61,14 @@ object MiddlewareSpec {
   def addHeader(key: String, value: String): MiddlewareSpec[Unit, Unit] =
     MiddlewareSpec(HttpCodec.empty, HeaderCodec.header(key, TextCodec.constant(value)))
 
-  val auth: MiddlewareSpec[Auth.Credentials, Unit] =
+  def basicAuth(u: String, p: String): MiddlewareSpec[Auth.Credentials, Unit] =
```

Comment:
I don't suggest we change this. In general, authentication is not done by passing hard-coded username and passpord, but by looking up information in a database (or at least config). I think we should revert this.

Also, this middleware spec is doing more than just being a `spec`: it is actually providing an implementation. That's an anti-pattern. The only point of a middleware spec is to define the (1) requirements on the request, and (2) the addendums to the response. No middleware logic should appear inside a `spec`. It must only ever appear inside the implementation for the middleware.

Reply:
>In general, authentication is not done by passing hard-coded username and passpord, but by looking up information in a database (or at least config). I think we should revert this.


I did this just trying to reimplement an existing `basicAuth` in the library.

Have a look at `Auth.scala` 

```scala
  /**
   * Creates a middleware for basic authentication that checks if the
   * credentials are same as the ones given
  final def basicAuth(u: String, p: String)(implicit trace: Trace): HttpMiddleware[Any, Nothing] =
    basicAuth { case credentials => (credentials.uname == u) && (credentials.upassword == p) }
```

But this can question our overly powerful `transformOrFail` in `MiddlewareSpec`.
The way it can get abused many `middlwareIn` may get _implemented_ as part of transformOrFail.

But may be its a smaller Pam

Reply:
All right, I suppose that's intended for a "quick & dirty" REST API. We can have that in the middleware, but let's leave the existing one there, and implement the new one in terms of the existing one. In any case, the logic should not be contained in the middleware.

I think you are right that `TransformOrFail` is maybe too powerful because it will encourage users to perform checks inside the specification, rather than inside the implementation. I think we are only using `transform` now for all the headers, so maybe we can get away with simplifying it.